src/main/java/com/iexec/blockchain/tool/QueueService.java

```diff
@@ -1,21 +1,136 @@
 package com.iexec.blockchain.tool;
 
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.stereotype.Service;
 
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
+import java.util.Objects;
+import java.util.concurrent.*;
 
+/**
+ * Execute {@link Runnable}s as they arrive.
+ * Some {@link Runnable}s can have higher priority than others, so they can be treated before the others.
+ */
+@Slf4j
 @Service
 public class QueueService {
-
     private final ExecutorService executorService;
+    private final PriorityBlockingQueue<BlockchainAction> queue = new PriorityBlockingQueue<>();
+
+    private CompletableFuture<Void> taskExecutor;
 
     public QueueService() {
         executorService = Executors.newFixedThreadPool(1);
     }
 
-    public void runAsync(Runnable runnable) {
-        CompletableFuture.runAsync(runnable, executorService);
+    /**
+     * Scheduled method execution.
+     * If the {@link QueueService#taskExecutor} isn't running, start a new thread.
+     * Otherwise, do nothing.
+     */
+    @Scheduled(fixedRate = 30000)
+    private void startAsyncTasksExecution() {
+        if (taskExecutor != null && !taskExecutor.isDone()) {
+            return;
+        }
+        
+        taskExecutor = CompletableFuture.runAsync(this::executeTasks);
+    }
+
+    /**
+     * Execute {@link Runnable}s as they come.
+     * At each occurrence, execute the first task in the queue or wait for a new one to appear.
+     * The first task is the task with the highest priority that is in the queue for the longer time.
+     */
+    void executeTasks() {
+        while (Thread.currentThread().isAlive()) {
+            executeFirstTask();
+        }
+    }
+
+    /**
+     * Execute the first task in the queue or wait for a new one to appear.
+     * The first task is the task with the highest priority that is in the queue for the longer time.
+     */
+    private void executeFirstTask() {
+        try {
+            // Wait until a new task is available.
+            Runnable runnable = queue.take().getRunnable();
+            // Execute a task and wait for its completion.
+            executorService.submit(runnable).get();
+        } catch (InterruptedException e) {
+            log.error("Task thread got interrupted.", e);
+            Thread.currentThread().interrupt();
+        } catch (RuntimeException e) {
+            log.error("An error occurred while waiting for new tasks.", e);
+        } catch (ExecutionException e) {
```

Comment:
We should catch all types of Throwable at the end to make sure the thread does not exit.